src/System.Private.ServiceModel/tests/Stress/SharedPoolsOfWCFObjects/PoolOfThings.cs

```diff
@@ -0,0 +1,374 @@
+            }
+            else
+            {
+                // we are the one who replaced the old _thePool - we're responsible for cleaning it up
+                DestroyAllPooledInstanciesImpl(oldPool, _destroyInstance);
+            }
+        }
+
+        public IEnumerable<T> GetAllPooledInstances()
+        {
+            return new AllPooledInstancesCollection(this);
+        }
+
+        #region helpers
+
+        private class AllPooledInstancesCollection : IEnumerable<T>
+        {
+            private PoolOfThings<T> _thePool;
+            public AllPooledInstancesCollection(PoolOfThings<T> thePool)
+            {
+                _thePool = thePool;
+            }
+
+            public IEnumerator<T> GetEnumerator()
+            {
+                for (int i = 0; i < _thePool._thePool.Length; i++)
+                {
+                    var instance = _thePool.GetPooledInstance(i);
+                    // skip nulls 
+                    if (instance != default(T))
+                    {
+                        yield return instance;
+                    }
+                }
+            }
+
+            IEnumerator IEnumerable.GetEnumerator()
+            {
+                return GetEnumerator();
+            }
+        }
+
+        private T GetPooledInstance(int i)
+        {
+            var instance = _thePool[i];
+            while (instance == null)
+            {
+                //try
+                {
+                    instance = _createInstance();
+                    if (instance == null)
+                    {
+                        break;
+                    }
+                    if (Interlocked.CompareExchange<T>(ref _thePool[i], instance, null) != null)
+                    {
+                        _destroyInstance(instance);
+                        instance = _thePool[i];
+                        // Somebody stored the new instance before us - dispose ours and re-read theirs
+                        // Since the whole pool can also be replaced we can still get a null from _thePool[index]
+                        // so we retry while (instance == null)
+                        // Note that _destroyInstance(instance) should not throw since nobody uses it yet
+
+                        //Interlocked.Increment(ref _numPoolSetRetries);
+                    }
+                }
+                //catch (Exception)
+                {
+                    //    break;
+                }
+            }
+            return instance;
+        }
+
+        private void DestroyAllPooledInstanciesImpl(T[] instancesPool, Action<T> destroyInstance)
+        {
+            foreach (var instance in instancesPool)
+            {
+                if (instance != null)
+                {
+                    try
+                    {
+                        //Interlocked.Increment(ref _stats.FactoriesClosed);
+                        destroyInstance(instance);
```

Comment:
For future consideration -- and I'll stop commenting on it.  These commented out Interlocked statements should either be removed or placed in some sort of conditionally compiled code (for debugging?)